File: patterns/sliding_window/minimumWindowSubstring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        
        need = {}
        for c in t:
            need[c] = need.get(c, 0) + 1
        
        window = {}
        l = 0
        have = 0
        needCount = len(need)
        length = float('inf')
        res = [-1,-1]

        for r in range(len(s)):
            c = s[r]
            window[c] = window.get(c, 0) + 1

            if c in need and window[c] == need[c]:
                have += 1
            
            while have == needCount:
                if (r - l + 1) < length:
                    res = [l,r]
                    length = r - l + 1
                
                window[s[l]] -= 1
                if s[l] in need and window[s[l]] < need[s[l]]:
                    have -= 1
                l += 1
        
        l, r = res
        return s[l:r+1] if length != float('inf') else ""

File: data_structures/strings/minimum_window_substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.size() < t.size()) return "";
        unordered_map<char, int> win;
        unordered_map<char, int> count;

        for (auto& c : t){
            count[c]++;
        }

        int have = 0;
        int need = count.size();

        int resL = INT_MAX;
        int res[2] = {-1,-1};

        int l = 0;

        for (int r = 0; r < s.size(); ++r){
            char c = s[r];
            win[c]++;

            if (win[c] == count[c]) have++;

            while (have == need){
                if (r-l+1 < resL){
                    res[0] = l;
                    res[1] = r;
                    resL = r - l + 1;
                }
                win[s[l]]--;
                if (win[s[l]] < count[s[l]]) have--;
                l++;
            }
        }

        return resL != INT_MAX ? s.substr(res[0], resL) : "";
    }
};
